### ISSUE TITLE
Accept a one-row-per-id DataFrame for `static_features`

### ISSUE THREAD

**Issue by simonez-tuidi** [Nixtla/mlforecast#702]
### Description

`static_features` today is `Optional[List[str]]`: a list of column *names* that
must already be materialized inside the long-format training `df`, repeated on
every timestamp of every series. This request proposes letting `static_features`
also accept a **DataFrame** (pandas or polars) holding one row per `id_col` and
the static columns as regular columns. When a DataFrame is passed, the caller no
longer needs to broadcast those values across every row of the main series
data — mlforecast reads them once, per id, from the smaller frame.

```python
# Today: brand/category/store_size are repeated on every one of the ~3M rows
# of `series` (one row per id per timestamp).
fcst.fit(series, static_features=["brand", "category", "store_size"])

# Proposed: pass the same information as a 3-column-per-id lookup table
# instead of asking the caller to broadcast it into every timestamp first.
statics = series[["unique_id", "brand", "category", "store_size"]].drop_duplicates("unique_id")
fcst.fit(series, static_features=statics)
```

## Motivation

Static features (things like a store's region, a product's category, a
customer's signup cohort) are constant for the lifetime of a series. mlforecast
already treats them this way internally: `TimeSeries._fit` takes exactly one
row per id from the incoming `df` (the last observation of each series, see
`core.py:634-649`) and stores the result as `self.static_features_`, a
dataframe with **one row per id** that is reused for every subsequent
`predict`/`update` call (`core.py:1320-1328`, `core.py:1539-1547`,
`core.py:1827-1859`).

In other words: the library's internal representation of static features is
*already* one-row-per-id. The current API just forces the caller to build the
wasteful, duplicated representation first, purely so the library can collapse
it right back down. For a caller who *has* their static attributes in a
natural one-row-per-id table already (a product catalog, a store master table,
a customer dimension table) — which is the common case, since that's usually
how this data is stored/queried in the first place — today's API requires an
explicit join/broadcast into the time-indexed data before calling mlforecast,
which:

- **Wastes memory proportional to series length.** A categorical/string static
  column repeated across a series with thousands of timestamps costs
  `n_timestamps` times what it needs to (pandas doesn't always dedupe
  string/object columns across rows the way a `Categorical` dtype would;
  polars is better here, but the row-with-strings materialization itself is
  still `O(n_series x n_timestamps)` rather than `O(n_series)`).
- **Costs an explicit join/merge step** in the caller's own pipeline before
  every `fit`/`preprocess`/`cross_validation` call, adding complexity and a
  second place where id mismatches or stale joins can introduce bugs.
- **Repeats the "did this static value drift?" validation
  (`_static_feature_changes_over_time`, `core.py:639-648`) over every
  timestamp** of every series, when the check is conceptually about a table
  with one row per id.

## Current behavior (for reference)

- `static_features=None` (default): every column in `df` other than
  `time_col`/`target_col` (and pooled `partition_by` columns) is auto-detected
  as static, and mlforecast raises if any of those columns actually varies
  within a series (`core.py:600-648`).
- `static_features=[...]` (explicit list of names): same mechanism, but scoped
  to the given column names; `id_col` is added automatically if missing
  (`core.py:613-623`).
- In both cases, the *values* are still expected to live in `df`, one
  (identical) copy per timestamp.
- The parameter is threaded through practically every public entry point that
  (re)builds series state: `MLForecast.preprocess`, `MLForecast.fit`,
  `MLForecast.history_warmup`, `MLForecast.cross_validation`,
  `TimeSeries._fit`/`fit_transform`/`history_warmup`,
  `DistributedMLForecast.fit`/`preprocess`, `LightGBMCV.fit`/`setup`, and
  `mlforecast.optimization.mlforecast_objective`.

## Proposed solution

Extend the type of `static_features` from `Optional[List[str]]` to
`Optional[Union[List[str], DataFrame]]` (`DataFrame` = the pandas/polars union
already used throughout the codebase via `utilsforecast.compat.DataFrame`).

When a DataFrame is passed:

1. It must contain `id_col` plus one or more other columns (the static
   feature values), with **exactly one row per id**.
2. The main training/update `df` is **not** required to contain those static
   columns at all — that's the entire point. If it does contain a column that
   also appears in the static DataFrame (other than `id_col`), that's treated
   as an ambiguous/conflicting specification and rejected with a clear error
   (see the implementation plan for the exact rule).
3. mlforecast joins the static DataFrame against the ids observed in `df` (in
   `self.uids` order) to build `self.static_features_` directly — skipping the
   per-timestamp "does this change over time" check entirely, since there's
   only one value per id to begin with.
4. Everything downstream that already consumes `self.static_features_` (the
   collapsed, one-row-per-id frame) is unaffected in behavior, because it
   already expects exactly this shape today — see the implementation plan for
   the (short) list of call sites that touch the *raw* `static_features`
   argument rather than the collapsed `static_features_` attribute, which are
   the only places that need branching logic.

This keeps `static_features` as a single, familiar parameter rather than
introducing a brand-new one (e.g. a separate `static_df=` argument), which:

- Requires no new parameter threading through the ~10 public methods that
  already forward `static_features`.
- Reads naturally: "these are my static features" — as either their *names*
  (look them up in `df`) or their *names-and-values* (here's the table).
- Is fully backward compatible: existing `None`/`List[str]` behavior is
  untouched.

### Example

```python
import pandas as pd
from mlforecast import MLForecast
from mlforecast.lag_transforms import ExpandingMean

series = pd.DataFrame({
    "unique_id": [...],   # long format, many rows per id
    "ds": [...],
    "y": [...],
})
store_attributes = pd.DataFrame({
    "unique_id": ["store_1", "store_2", ...],  # one row per id
    "region": ["west", "east", ...],
    "store_size_sqft": [12000, 8500, ...],
})

fcst = MLForecast(
    models=[...],
    freq="D",
    lag_transforms={7: [ExpandingMean()]},
)
fcst.fit(series, static_features=store_attributes)
fcst.predict(14)  # region / store_size_sqft come from store_attributes, not `series`
```

## Scope

In scope for v1:

- `TimeSeries._fit` / `fit_transform` / `history_warmup` (core mechanism).
- `MLForecast.preprocess` / `fit` / `history_warmup` / `cross_validation`
  (type-hint + validation updates; these already forward to `TimeSeries`
  unchanged).
- `TimeSeries.update` / `MLForecast.update`: sourcing static values for
  **brand-new** ids introduced mid-stream (not present at `fit` time). See the
  implementation plan for the proposed mechanism (a retained lookup registry
  plus an optional argument for truly new ids).
- `DistributedMLForecast` (dask/ray/spark via fugue): broadcasting the (small,
  one-row-per-id) static DataFrame to every partition.
- `LightGBMCV` and `mlforecast.optimization.mlforecast_objective`: passthrough
  support and categorical-feature detection (`optimization.py:23-43`, which
  currently assumes a list of names).

Out of scope / explicitly deferred:

- Accepting a *lazy* frame (polars `LazyFrame`) — the static table is expected
  to be small, so eager is fine for v1.
- Allowing the static DataFrame to be re-supplied or changed on every
  `predict`/`cross_validation` call independent of `fit` — for v1 it's fixed
  at `fit`/`preprocess`/`history_warmup` time, matching how `static_features_`
  already works today.
- Mixing "some static features via list-of-names-in-df" and "some via the new
  DataFrame" in the same call. v1 is one or the other, not both, to keep the
  validation and error messages unambiguous.

## Alternatives considered

1. **New parameter, e.g. `static_df=`.** Rejected: doubles the parameter
   surface across every method that currently accepts `static_features`, and
   raises questions about precedence when both are given. Reusing
   `static_features` with a richer type keeps one mental model: "how do I
   tell mlforecast about my static columns," with a value that's either a
   column-name list or the names-and-values table.
2. **Status quo: require the user to join the static table onto `df`
   themselves.** This is exactly the workflow this request wants to avoid;
   it's always been *possible*, that's not the problem — it's the fact that
   it's the *only* option today, and it wastes memory for no functional
   benefit since the library immediately collapses it back down.
3. **Accept a `dict[id, dict[col, value]]` mapping.** Rejected: awkward for
   many static columns, doesn't compose with polars, and offers no advantage
   over a DataFrame with one row per id — which is also what most users
   already have on hand (e.g. a dimension table read straight from a
   warehouse query).

## Open questions for maintainers

- **Overlap-column policy.** Should a column present in both the main `df`
  and the static DataFrame (other than `id_col`) hard-error, or should the
  static DataFrame silently take precedence? This proposal defaults to a hard
  error (fail fast, avoid silent surprises) — see the implementation plan.
- **New ids via `update()`.** Since the main `update()` `df` is not expected to
  carry static columns either, how should statics for a brand-new id be
  supplied? The implementation plan proposes (a) retaining any unmatched rows
  from the original static DataFrame as a lookup registry for ids that show up
  later, and (b) an optional `new_static_features` argument on `update()` for
  ids not already present in that registry. Feedback welcome on whether both
  are needed for v1 or if the registry alone covers the common case.
- **Distributed engines.** Fugue-based `DistributedMLForecast` shards `df` by
  id across partitions/workers. The static DataFrame is small and needs to
  reach every worker; the implementation plan proposes broadcasting it via
  fugue's `params`, the same mechanism already used to pass `static_features`
  (as a list) to `_preprocess_partition` today (`distributed/forecast.py:160-239`).
  Confirm this is an acceptable broadcast strategy at the expected static-table
  sizes (single-digit MBs, not GBs).

## Related code references (as of commit 5607b92)

- `mlforecast/core.py:540-756` — `TimeSeries._fit` (where `static_features_` is
  built and validated).
- `mlforecast/core.py:1234-1305` — `TimeSeries.history_warmup` (same
  construction path, feature-free).
- `mlforecast/core.py:1320-1328`, `1539-1547` — where `static_features_` is
  concatenated with computed features at predict time.
- `mlforecast/core.py:1827-1859` — `_maybe_subset`, which indexes
  `static_features_` by row position aligned to `self.uids`.
- `mlforecast/core.py:2019-2113` — `TimeSeries.update`, including the
  new-series static extraction at `2075-2088` that currently reads from the
  `update()` `df`.
- `mlforecast/forecast.py:273-294` — `MLForecast._get_dynamic_exog_cols`, a
  second, forecast.py-local implementation of the "what's static vs.
  dynamic-exogenous" split that also assumes a list of names.
- `mlforecast/distributed/forecast.py:160-239`, `960-1047` — per-partition
  preprocessing and post-fit combination of `static_features_` across
  partitions.
- `mlforecast/optimization.py:23-43` — categorical-static-feature detection
  for LightGBM's `categorical_feature`, assumes `static_features: List[str]`.
- `mlforecast/lgb_cv.py:170-245`, `415-500` — `LightGBMCV` passthrough.

**Comment by jmoralez** [comment]
The static features are used by the models during training, how would that work?

**Comment by simonez-tuidi** [comment]
@jmoralez 

If they are included as actual training features then they could be kept as they are, but there are a lot of applications where those static features are only used as keys for the `groupby` parameter and therefore are not needed in the training set at all, they're actually dropped from the `preprocess` output as auxiliary columns by default

**Comment by jmoralez** [comment]
So they wouldn't be static features but rather extra IDs, I think they should be named accordingly.
